File: unittest/scripts/auto/js_adminapi_async/scripts/simple_plain.js
//@ {VER(>=8.0.11)}

// Plain replicaset setup test, use as a template for other tests that check
// a specific feature/aspect across the whole API

//@<> Setup
testutil.deployRawSandbox(__mysql_sandbox_port1, "root");
testutil.snapshotSandboxConf(__mysql_sandbox_port1);
testutil.deploySandbox(__mysql_sandbox_port2, "root");
testutil.snapshotSandboxConf(__mysql_sandbox_port2);
testutil.deploySandbox(__mysql_sandbox_port3, "root");
testutil.snapshotSandboxConf(__mysql_sandbox_port3);

shell.options.useWizards = false;

//@ configureReplicaSetInstance + create admin user
dba.configureReplicaSetInstance(__sandbox_uri1, {clusterAdmin:"admin", clusterAdminPassword:"bla"});
testutil.restartSandbox(__mysql_sandbox_port1);

//@ configureReplicaSetInstance
dba.configureReplicaSetInstance(__sandbox_uri2, {clusterAdmin:"admin", clusterAdminPassword:"bla"});
dba.configureReplicaSetInstance(__sandbox_uri3, {clusterAdmin:"admin", clusterAdminPassword:"bla"});

//@ createReplicaSet
shell.connect(__sandbox_uri1);

rs = dba.createReplicaSet("myrs");

//@ status
rs.status();

//@ disconnect
rs.disconnect();

//@ getReplicaSet
rs = dba.getReplicaSet();

//@ addInstance
rs.addInstance(__sandbox_uri2, {recoveryMethod:'incremental'});

rs.addInstance(__sandbox_uri3, {recoveryMethod:'incremental'});

//@ removeInstance
rs.removeInstance(__sandbox_uri2);

rs.addInstance(__sandbox_uri2, {recoveryMethod:'incremental'});

//@ setPrimaryInstance
rs.setPrimaryInstance(__sandbox_uri3);

//@ forcePrimaryInstance (prepare)
testutil.stopSandbox(__mysql_sandbox_port3);
rs = dba.getReplicaSet();

rs.status();

//@ forcePrimaryInstance
rs.forcePrimaryInstance(__sandbox_uri1);

//@ rejoinInstance
testutil.startSandbox(__mysql_sandbox_port3);

rs.rejoinInstance(__sandbox_uri3);

//@ listRouters
cluster_id = session.runSql("SELECT cluster_id FROM mysql_innodb_cluster_metadata.clusters").fetchOne()[0];
session.runSql("INSERT mysql_innodb_cluster_metadata.routers VALUES (DEFAULT, 'system', 'mysqlrouter', 'routerhost1', '8.0.18', '2019-01-01 11:22:33', NULL, ?, NULL)", [cluster_id]);
session.runSql("INSERT mysql_innodb_cluster_metadata.routers VALUES (DEFAULT, 'system', 'mysqlrouter', 'routerhost2', '8.0.18', '2019-01-01 11:22:33', NULL, ?, NULL)", [cluster_id]);

rs.listRouters();

//@ removeRouterMetadata
rs.removeRouterMetadata("routerhost1::system");

rs.listRouters();

//@ createReplicaSet(adopt)
session.runSql("DROP SCHEMA mysql_innodb_cluster_metadata");
//BREAK
rs = dba.createReplicaSet("adopted", {adoptFromAR:true});

rs.status();

//@<> Cleanup
testutil.destroySandbox(__mysql_sandbox_port1);
testutil.destroySandbox(__mysql_sandbox_port2);
testutil.destroySandbox(__mysql_sandbox_port3);

File: unittest/scripts/auto/py_adminapi_async/scripts/simple_plain.py
#@ {VER(>=8.0.11)}

# Plain replicaset setup test, use as a template for other tests that check
# a specific feature/aspect across the whole API

#@<> Setup
testutil.deploy_raw_sandbox(port=__mysql_sandbox_port1, rootpass="root")
testutil.snapshot_sandbox_conf(port=__mysql_sandbox_port1)
testutil.deploy_sandbox(port=__mysql_sandbox_port2, rootpass="root")
testutil.snapshot_sandbox_conf(port=__mysql_sandbox_port2)
testutil.deploy_sandbox(port=__mysql_sandbox_port3, rootpass="root")
testutil.snapshot_sandbox_conf(port=__mysql_sandbox_port3)

shell.options.useWizards = False

#@ configure_replica_set_instance + create admin user
dba.configure_replica_set_instance(instance=__sandbox_uri1, options={"clusterAdmin":"admin", "clusterAdminPassword":"bla"})
testutil.restart_sandbox(port=__mysql_sandbox_port1)

#@ configure_replica_set_instance
dba.configure_replica_set_instance(instance=__sandbox_uri2, options={"clusterAdmin":"admin", "clusterAdminPassword":"bla"})
dba.configure_replica_set_instance(instance=__sandbox_uri3, options={"clusterAdmin":"admin", "clusterAdminPassword":"bla"})

#@ create_replica_set
shell.connect(connectionData=__sandbox_uri1)

rs = dba.create_replica_set(name="myrs")

#@ status
rs.status()

#@ disconnect
rs.disconnect()

#@ get_replica_set
rs = dba.get_replica_set()

#@<> add_instance using clone recovery {VER(>=8.0.17)}
rs.add_instance(instance=__sandbox_uri2, options={"recoveryMethod":'clone'})

#@<> add_instance using incremental recovery {VER(<8.0.17)}
rs.add_instance(instance=__sandbox_uri2, options={"recoveryMethod":'incremental'})

#@<> add_instance 3
rs.add_instance(instance=__sandbox_uri3, options={"recoveryMethod":'incremental'})

#@ remove_instance
rs.remove_instance(instance=__sandbox_uri2)

rs.add_instance(instance=__sandbox_uri2, options={"recoveryMethod":'incremental'})

#@ set_primary_instance
rs.set_primary_instance(instance=__sandbox_uri3)

#@ force_primary_instance (prepare)
testutil.stop_sandbox(port=__mysql_sandbox_port3, options={"wait":1})
rs = dba.get_replica_set()

rs.status()

#@ force_primary_instance
rs.force_primary_instance(instance=__sandbox_uri1)

# rejoinInstance
testutil.start_sandbox(port=__mysql_sandbox_port3)

# XXX rs.rejoinInstance(__sandbox_uri3)

#@ list_routers
cluster_id = session.run_sql("SELECT cluster_id FROM mysql_innodb_cluster_metadata.clusters").fetch_one()[0]
session.run_sql("INSERT mysql_innodb_cluster_metadata.routers VALUES (DEFAULT, 'system', 'mysqlrouter', 'routerhost1', '8.0.18', '2019-01-01 11:22:33', NULL, ?, NULL, NULL)", [cluster_id])
session.run_sql("INSERT mysql_innodb_cluster_metadata.routers VALUES (DEFAULT, 'system', 'mysqlrouter', 'routerhost2', '8.0.18', '2019-01-01 11:22:33', NULL, ?, NULL, NULL)", [cluster_id])

rs.list_routers()

#@ remove_router_metadata
rs.remove_router_metadata(routerDef="routerhost1::system")

rs.list_routers()

#@ create_replica_set(adopt)
session.run_sql("DROP SCHEMA mysql_innodb_cluster_metadata")
testutil.wait_member_transactions(__mysql_sandbox_port2, __mysql_sandbox_port1);

rs = dba.create_replica_set(name="adopted", options={"adoptFromAR":True})

#@<> create_replica_set(adopt) - status
rs.status()

rs.disconnect()

#@<> Cleanup
testutil.destroy_sandbox(port=__mysql_sandbox_port1)
testutil.destroy_sandbox(port=__mysql_sandbox_port2)
testutil.destroy_sandbox(port=__mysql_sandbox_port3)
